src/noit_lmdb_tools.c

```diff
@@ -0,0 +1,308 @@
+  if (key_len) {
+    memcpy(current_location, key, key_len);
+    current_location += key_len;
+  }
+  /* Verify that we copied in what we expected to - this is checking for programming
+   * errors */
+  mtevAssert((current_location - size) == toRet);
+  if (size_out) {
+    *size_out = size;
+  }
+  return toRet;
+}
+inline char *
+noit_lmdb_make_check_key_for_iterating(uuid_t id, size_t *size_out) {
+  char *toRet = (char *)malloc(UUID_SIZE);
+  memcpy(toRet, id, UUID_SIZE);
+  if (size_out) {
+    *size_out = UUID_SIZE;
+  }
+  return toRet;
+}
+noit_lmdb_check_data_t *noit_lmdb_check_data_from_key(char *key) {
+  noit_lmdb_check_data_t *toRet = NULL;
+  size_t current_location = 0;
+  if (!key) {
+    return toRet;
+  }
+  toRet = (noit_lmdb_check_data_t *)calloc(1, sizeof(noit_lmdb_check_data_t));
+  memcpy(&toRet->id, key, UUID_SIZE);
+  current_location += UUID_SIZE;
+  memcpy(&toRet->type, key + current_location, sizeof(char));
+  current_location += sizeof(char);
+  memcpy(&toRet->ns_len, key + current_location, sizeof(unsigned short));
+  toRet->ns_len = ntohs(toRet->ns_len);
+  current_location += sizeof(unsigned short);
+  if (toRet->ns_len) {
+    toRet->ns = (char *)calloc(1, toRet->ns_len + 1);
+    memcpy(toRet->ns, key + current_location, toRet->ns_len);
+    current_location += toRet->ns_len;
+  }
+  memcpy(&toRet->key_len, key + current_location, sizeof(unsigned short));
+  toRet->key_len = ntohs(toRet->key_len);
+  current_location += sizeof(unsigned short);
+  if (toRet->key_len) {
+    toRet->key = (char *)calloc(1, toRet->key_len + 1);
+    memcpy(toRet->key, key + current_location, toRet->key_len);
+    current_location += toRet->ns_len;
+  }
+  return toRet;
+}
+
+void noit_lmdb_free_check_data(noit_lmdb_check_data_t *data) {
+  if (data) {
+    free(data->ns);
+    free(data->key);
+    free(data);
+  }
+}
+
+noit_lmdb_instance_t *noit_lmdb_tools_open_instance(char *path)
+{
+  int rc;
+  MDB_env *env;
+
+  mtevAssert(lmdb_instance_mkdir(path));
+
+  rc = mdb_env_create(&env);
+  if (rc != 0) {
+    errno = rc;
+    return NULL;
+  }
+
+  rc = mdb_env_set_maxreaders(env, 1024);
+  if (rc != 0) {
+    errno = rc;
+    mdb_env_close(env);
+    return NULL;
+  }
+
+  rc = mdb_env_open(env, path, 0, 0644);
+  if (rc != 0) {
+    errno = rc;
+    mdb_env_close(env);
+    return NULL;
+  }
+
+  MDB_txn *txn;
+  MDB_dbi dbi;
+  rc = mdb_txn_begin(env, NULL, 0, &txn);
+  if (rc != 0) {
+    errno = rc;
+    mdb_env_close(env);
+    return NULL;
+  }
+  rc = mdb_open(txn, NULL, MDB_CREATE, &dbi);
+  if (rc != 0) {
+    mdb_txn_abort(txn);
+    mdb_env_close(env);
+    errno = rc;
+    return NULL;
+  }
+  rc = mdb_txn_commit(txn);
+  if (rc != 0) {
+    mdb_txn_abort(txn);
+    mdb_env_close(env);
+    errno = rc;
+    return NULL;
+  }
+
+  noit_lmdb_instance_t *instance = (noit_lmdb_instance_t *)malloc(sizeof(noit_lmdb_instance_t));
+  instance->env = env;
+  instance->dbi = dbi;
+  ck_rwlock_init(&instance->lock);
```

Comment:
This should be a pthread_wrlock_t...the operations it surrounds are big enough that spinning would be detrimental

Reply:
Fixed - uses a `pthread_rwlock_t` now